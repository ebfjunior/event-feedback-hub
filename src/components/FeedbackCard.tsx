import { Star } from 'lucide-react';
import { cn } from '@/lib/utils';
import type { FeedbackItem } from '@/types/api';

export type FeedbackCardProps = {
  feedback: FeedbackItem;
  highlight?: boolean;
};

function Stars({ rating }: { rating: number }) {
  return (
    <div className="flex items-center gap-1" aria-label={`${rating} out of 5 stars`}>
      {Array.from({ length: 5 }).map((_, i) => (
        <Star
          key={i}
          className={cn('h-4 w-4', i < rating ? 'fill-yellow-400 stroke-yellow-400' : 'stroke-muted-foreground')}
        />
      ))}
    </div>
  );
}

export function FeedbackCard({ feedback, highlight }: FeedbackCardProps) {
  const created = new Date(feedback.created_at);
  const timeAgo = new Intl.RelativeTimeFormat(undefined, { numeric: 'auto' });
  const diffMs = Date.now() - created.getTime();
  const diffMin = Math.round(diffMs / 60000);
  const rtf = diffMin < 60 ? timeAgo.format(-diffMin, 'minute') : timeAgo.format(-Math.round(diffMin / 60), 'hour');

  return (
    <div className={cn('rounded-md border bg-card/60 p-3 shadow-sm', highlight && 'ring-1 ring-yellow-400')}> 
      <div className="mb-1 flex items-center justify-between text-xs text-muted-foreground">
        <div className="truncate font-medium text-primary/80">{feedback.event_name}</div>
        <div>{rtf}</div>
      </div>
      <div className="mb-2 text-sm text-foreground/90">{feedback.text}</div>
      <div className="flex items-center justify-between">
        <div className="text-xs text-muted-foreground">Anonymous</div>
        <Stars rating={feedback.rating} />
      </div>
    </div>
  );
}
