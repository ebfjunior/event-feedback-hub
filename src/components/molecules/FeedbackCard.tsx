import Link from 'next/link';
import type { FeedbackItem } from '@/types/api';
import { cn } from '@/lib/utils';
import { Card, CardContent, CardDescription, CardHeader, CardAction } from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { Stars } from '@/components/atoms/Stars';

export type FeedbackCardProps = {
  feedback: FeedbackItem;
  highlight?: boolean;
};

export function FeedbackCard({ feedback, highlight }: FeedbackCardProps) {
  const created = new Date(feedback.created_at);
  const timeAgo = new Intl.RelativeTimeFormat(undefined, { numeric: 'auto' });
  const diffMs = Date.now() - created.getTime();
  const diffMin = Math.round(diffMs / 60000);
  const rtf = diffMin < 60 ? timeAgo.format(-diffMin, 'minute') : timeAgo.format(-Math.round(diffMin / 60), 'hour');

  return (
    <Card className={cn('bg-card/60 shadow-sm', highlight && 'ring-1 ring-yellow-400')}>
      <CardHeader className="gap-1.5">
        <Badge asChild variant="secondary" className="truncate max-w-full">
          <Link href={`/events/${feedback.event_id}`} aria-label={`Open event ${feedback.event_name}`}>
            {feedback.event_name}
          </Link>
        </Badge>
        <CardDescription className="text-xs">{rtf}</CardDescription>
        <CardAction>
          <Stars rating={feedback.rating} />
        </CardAction>
      </CardHeader>
      <CardContent className="pt-0">
        <div className="text-sm text-foreground/90">{feedback.text}</div>
        <div className="mt-3 text-xs text-muted-foreground">Anonymous</div>
      </CardContent>
    </Card>
  );
}


