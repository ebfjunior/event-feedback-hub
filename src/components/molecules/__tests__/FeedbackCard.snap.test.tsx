import React from 'react';
import { render } from '@testing-library/react';
import { FeedbackCard } from '../FeedbackCard';

vi.mock('next/link', () => ({ default: (props: any) => <a {...props} /> }));
vi.mock('@/components/atoms/Stars', () => ({
	Stars: ({ rating }: { rating: number }) => <div data-stars>{rating}</div>,
}));

const base = {
	id: '1',
	event_id: 'e1',
	event_name: 'Event 1',
	rating: 4,
	text: 'Great event',
	created_at: new Date('2024-01-01T00:10:00Z').toISOString(),
};

describe('FeedbackCard (snapshot)', () => {
	it('renders highlighted', () => {
		const { container } = render(<FeedbackCard feedback={base} highlight />);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('renders non-highlighted', () => {
		const { container } = render(<FeedbackCard feedback={base} />);
		expect(container.firstChild).toMatchSnapshot();
	});
});


