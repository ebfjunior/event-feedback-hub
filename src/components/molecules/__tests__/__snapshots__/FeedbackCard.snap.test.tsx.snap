// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FeedbackCard (snapshot) > renders highlighted 1`] = `
<div
  class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 bg-card/60 shadow-sm ring-1 ring-yellow-400"
  data-slot="card"
>
  <div
    class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 gap-1.5"
    data-slot="card-header"
  >
    <a
      aria-label="Open event Event 1"
      class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 truncate max-w-full"
      data-slot="badge"
      href="/events/e1"
    >
      Event 1
    </a>
    <div
      class="text-muted-foreground text-xs"
      data-slot="card-description"
    >
      in 10 minutes
    </div>
    <div
      class="col-start-2 row-span-2 row-start-1 self-start justify-self-end"
      data-slot="card-action"
    >
      <div
        data-stars="true"
      >
        4
      </div>
    </div>
  </div>
  <div
    class="px-6 pt-0"
    data-slot="card-content"
  >
    <div
      class="text-sm text-foreground/90"
    >
      Great event
    </div>
    <div
      class="mt-3 text-xs text-muted-foreground"
    >
      Anonymous
    </div>
  </div>
</div>
`;

exports[`FeedbackCard (snapshot) > renders non-highlighted 1`] = `
<div
  class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 bg-card/60 shadow-sm"
  data-slot="card"
>
  <div
    class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 gap-1.5"
    data-slot="card-header"
  >
    <a
      aria-label="Open event Event 1"
      class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 truncate max-w-full"
      data-slot="badge"
      href="/events/e1"
    >
      Event 1
    </a>
    <div
      class="text-muted-foreground text-xs"
      data-slot="card-description"
    >
      in 10 minutes
    </div>
    <div
      class="col-start-2 row-span-2 row-start-1 self-start justify-self-end"
      data-slot="card-action"
    >
      <div
        data-stars="true"
      >
        4
      </div>
    </div>
  </div>
  <div
    class="px-6 pt-0"
    data-slot="card-content"
  >
    <div
      class="text-sm text-foreground/90"
    >
      Great event
    </div>
    <div
      class="mt-3 text-xs text-muted-foreground"
    >
      Anonymous
    </div>
  </div>
</div>
`;
